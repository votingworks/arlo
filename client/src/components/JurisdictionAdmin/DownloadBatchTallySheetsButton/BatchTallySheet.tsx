import React, { ReactNode } from 'react'
import { Page, StyleSheet, Text, View } from '@react-pdf/renderer'
import { Style } from '@react-pdf/types'

import { blankLine } from '../../../utils/string'
import { ICandidate } from '../../../types'
import {
  PdfCheckbox,
  PdfDividerLine,
  PdfP,
  PdfHeading,
  PdfSignatureLine,
  PdfSubHeading,
  PdfTable,
  PdfTd,
  PdfTr,
} from '../../Atoms/pdf'

const styles = StyleSheet.create({
  page: {
    fontFamily: 'Helvetica', // @react-pdf comes pre-packaged with this font
    fontSize: 12,
    padding: 36,
  },
  pageSection: {
    marginBottom: 24,
  },
  bold: {
    fontFamily: 'Helvetica-Bold', // @react-pdf comes pre-packaged with this font
  },
  flexRow: {
    display: 'flex',
    flexDirection: 'row',
  },
  overflowHidden: {
    overflow: 'hidden',
  },
  withSmallBottomMargin: {
    marginBottom: 10,
  },
})

interface IPageSectionProps {
  children: ReactNode
  style?: Style
}

const PageSection = ({ children, style }: IPageSectionProps): JSX.Element => {
  return (
    <View
      style={{
        ...styles.pageSection,
        ...style,
      }}
    >
      {children}
    </View>
  )
}

interface IProps {
  auditBoardName: string
  batchName: string
  choices: ICandidate[]
  jurisdictionName: string
}

const BatchTallySheet = ({
  auditBoardName,
  batchName,
  choices,
  jurisdictionName,
}: IProps): JSX.Element => {
  return (
    <Page style={styles.page} size="LETTER">
      <PageSection>
        <PdfHeading>Audit Board Batch Tally Sheet</PdfHeading>
        <PdfSubHeading>
          Batch Name: <Text style={styles.bold}>{batchName}</Text>
        </PdfSubHeading>
      </PageSection>

      <PageSection>
        <PdfP>Jurisdiction: {jurisdictionName}</PdfP>
        <PdfP>Audit Board: {auditBoardName}</PdfP>
        <PdfP lastInSection>Batch Type (Optional): {blankLine(20)}</PdfP>
      </PageSection>

      <PageSection style={styles.flexRow}>
        <Text>
          Was the container sealed when received by the audit board?&nbsp;&nbsp;
        </Text>
        <PdfCheckbox fontSize={styles.page.fontSize} />
        <Text>&nbsp;Yes</Text>
      </PageSection>

      <PageSection>
        <PdfTable>
          <PdfTr>
            <PdfTd>
              <Text style={styles.bold}>Candidates/Choices</Text>
            </PdfTd>
            <PdfTd>
              <Text style={styles.bold}>Enter Stack Totals</Text>
            </PdfTd>
          </PdfTr>
          {choices.map(c => (
            <PdfTr key={c.id}>
              <PdfTd>
                <Text style={styles.overflowHidden}>{c.name}</Text>
              </PdfTd>
              {/* Blank, to be filled in by audit board */}
              <PdfTd />
            </PdfTr>
          ))}
        </PdfTable>
      </PageSection>

      <PageSection>
        <PdfP>
          When work is completed, return all ballots to the ballot container and
          seal the container.
        </PdfP>
        <View style={styles.flexRow}>
          <Text>
            Was the container resealed by the audit board?&nbsp;&nbsp;
          </Text>
          <PdfCheckbox fontSize={styles.page.fontSize} />
          <Text>&nbsp;Yes</Text>
        </View>
      </PageSection>

      <PageSection style={styles.flexRow}>
        <PdfSignatureLine
          label="(Audit Board Member)"
          style={{ marginRight: 10 }}
        />
        <PdfSignatureLine label="(Audit Board Member)" />
      </PageSection>

      <PageSection>
        <PdfDividerLine />
      </PageSection>

      <PageSection>
        <PdfP style={styles.bold}>Check-In/Out Station Steps:</PdfP>
        <View style={[styles.flexRow, styles.withSmallBottomMargin]}>
          <PdfCheckbox fontSize={styles.page.fontSize} />
          <Text>&nbsp;Recorded batch check-in</Text>
        </View>
        <View style={[styles.flexRow, styles.withSmallBottomMargin]}>
          <PdfCheckbox fontSize={styles.page.fontSize} />
          <Text>&nbsp;Entered tallies into Arlo</Text>
        </View>
        <PdfP lastInSection>
          {blankLine(5)} Initials of check-in/out station member
        </PdfP>
      </PageSection>
    </Page>
  )
}

export default BatchTallySheet
