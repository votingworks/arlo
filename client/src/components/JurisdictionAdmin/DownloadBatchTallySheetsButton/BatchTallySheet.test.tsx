import React from 'react'
import { render } from '@testing-library/react'

import BatchTallySheet from './BatchTallySheet'

jest.mock('@react-pdf/renderer', () => ({
  ...jest.requireActual('@react-pdf/renderer'),

  // Mock @react-pdf/renderer to generate HTML instead of PDF content for easier testing
  Document: jest.fn(({ children }) => <div>{children}</div>),
  Page: jest.fn(({ children }) => <div>{children}</div>),
  Text: jest.fn(({ children }) => <div>{children}</div>),
  View: jest.fn(({ children }) => <div>{children}</div>),
}))

test('BatchTallySheet renders expected PDF content', async () => {
  const { container } = render(
    <BatchTallySheet
      auditBoardName="Audit Board 1"
      batchName="Batch 1"
      choices={[
        { id: 'choice-1', name: 'Choice 1', numVotes: 0 },
        { id: 'choice-2', name: 'Choice 2', numVotes: 0 },
        { id: 'choice-3', name: 'Choice 3', numVotes: 0 },
      ]}
      jurisdictionName="Jurisdiction 1"
    />
  )
  expect(container).toMatchSnapshot()
})
