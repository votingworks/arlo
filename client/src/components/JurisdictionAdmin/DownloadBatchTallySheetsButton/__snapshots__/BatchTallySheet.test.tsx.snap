// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchTallySheet renders expected PDF content 1`] = `
<div>
  <div>
    <div>
      <div>
        Audit Board Batch Tally Sheet
      </div>
      <div>
        Batch Name: 
        <div>
          Batch 1
        </div>
      </div>
    </div>
    <div>
      <div>
        Jurisdiction: 
        Jurisdiction 1
      </div>
      <div>
        Audit Board: 
        Audit Board 1
      </div>
      <div>
        Batch Type (Optional): 
        ____________________
      </div>
    </div>
    <div>
      <div>
        Was the container sealed when received by the audit board?  
      </div>
      <div />
      <div>
         Yes
      </div>
    </div>
    <div>
      <div>
        <div>
          <div>
            <div>
              Candidates/Choices
            </div>
          </div>
          <div>
            <div>
              Enter Stack Totals
            </div>
          </div>
        </div>
        <div>
          <div>
            <div>
              Choice 1
            </div>
          </div>
          <div />
        </div>
        <div>
          <div>
            <div>
              Choice 2
            </div>
          </div>
          <div />
        </div>
        <div>
          <div>
            <div>
              Choice 3
            </div>
          </div>
          <div />
        </div>
      </div>
    </div>
    <div>
      <div>
        When work is completed, return all ballots to the ballot container and seal the container.
      </div>
      <div>
        <div>
          Was the container resealed by the audit board?  
        </div>
        <div />
        <div>
           Yes
        </div>
      </div>
    </div>
    <div>
      <div>
        <div>
          x
          ______________________________
        </div>
        <div>
          (Audit Board Member)
        </div>
      </div>
      <div>
        <div>
          x
          ______________________________
        </div>
        <div>
          (Audit Board Member)
        </div>
      </div>
    </div>
    <div>
      <div />
    </div>
    <div>
      <div>
        Check-In/Out Station Steps:
      </div>
      <div>
        <div />
        <div>
           Recorded batch check-in
        </div>
      </div>
      <div>
        <div />
        <div>
           Entered tallies into Arlo
        </div>
      </div>
      <div>
        _____
         Initials of check-in/out station member
      </div>
    </div>
  </div>
</div>
`;
